### ISSUE TITLE
When training multiple concepts at once, looks like only first gets trained

### ISSUE THREAD

**Issue by taleksan** [d8ahazard/sd_dreambooth_extension#743]
### Kindly read the entire form below and fill it out with the requested information.

**Please find the following lines in the console and paste them below. _If you do not provide this information, your
issue will be automatically closed._**

`
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Commit hash: 8850fc23b6e8a8e210bdfe4aade81516fb5770f3
Installing requirements for Web UI

#######################################################################################################
Initializing Dreambooth
If submitting an issue on github, please provide the below text for debugging purposes:

Python revision: 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Dreambooth revision: 39a937a47e074afa5508c0ceac53bd2c06cc33ca
SD-WebUI revision: 8850fc23b6e8a8e210bdfe4aade81516fb5770f3

Checking Dreambooth requirements...
[+] bitsandbytes version 0.35.0 installed.
[+] diffusers version 0.10.2 installed.
[+] transformers version 4.25.1 installed.
[+] xformers version 0.0.14.dev0 installed.
[+] torch version 1.12.1+cu113 installed.
[+] torchvision version 0.13.1+cu113 installed.

#######################################################################################################

Launching Web UI with arguments: --xformers
`

**Have you read the [Readme](https://github.com/d8ahazard/sd_dreambooth_extension#readme)?**
yes

**Have you completely restarted the stable-diffusion-webUI, not just reloaded the UI?**
yes

**Have you updated Dreambooth to the latest revision?**
yes - that's when the problem started 

**Have you updated the Stable-Diffusion-WebUI to the latest version?**
yes

**No, really. Please save us both some trouble and update the SD-WebUI and Extension and restart before posting this.
Reply 'OK' Below to acknowledge that you did this.**
OK

**Describe the bug**

The Concept 2 and Concept 3 seem to be ignored after I updated to latest webui and dreambooth. The model only trains Concept 1. Samples generated only reflect concept 1 even if the prompt provides is concept 2 or 3 instance.

**Provide logs**

If a crash has occurred, please provide the *entire* stack trace from the log, including the last few log messages *
before* the crash occurred.

**Environment**

What OS?
Win 11

If Windows - WSL or native?
native

What GPU are you using?
RTX 3090

**Screenshots/Config**
If the issue is specific to an error while training, please provide a screenshot of training parameters or the
db_config.json file from /models/dreambooth/MODELNAME/db_config.json
[db_config.json.txt](https://github.com/d8ahazard/sd_dreambooth_extension/files/10371247/db_config.json.txt)

**Comment by ExponentialML** [comment]
After entering your parameters into concept's 2 and/or 3, have you tried hitting the save button *first* before training?
There may be a bug where the individual concepts don't save automatically when hitting the train button.

**Comment by d8ahazard** [comment]
> After entering your parameters into concept's 2 and/or 3, have you tried hitting the save button _first_ before training? There may be a bug where the individual concepts don't save automatically when hitting the train button.

I believe that the majority of UI functions that require the config be saved again before training now have this behaviour fixed. I've even added a little popup box that warns of a possible overwrite with other settings.


It would be nice to se the saved db_config file, or at least the concepts_list portion to determine the issue.

**Comment by Drante177** [comment]
Since recently updating the Dreambooth Extension, I have been experiencing the same issue.

Taleksan does appear to have provided the db_config file with their original report, mine is not very different.

One odd thing I notice when starting training is that is repeats that it is using "Bucket 0", so it loads the first concept into Bucket 0, then when the second concept loads it says Bucket 0 too.

> Instance Bucket 0: Resolution (512, 512), Count: 22
> Class Bucket 0: Resolution (512, 512), Count: 499
> Instance Bucket 0: Resolution (512, 512), Count: 20
> Class Bucket 0: Resolution (512, 512), Count: 499
> Target Bucket 0: Resolution (512, 512), Count: 0

My third concept is empty, so I guess we expect that to be showing and having a count of zero.

The results of the training do seem to favour the first concept, but whether I use the token of the first concept or the second concept the output image is nearly identical while using the same seed.

**Comment by tsngo** [comment]
Same thing happens for me. When saving out preview images, the save_sample_prompt is correct but the image renders only the first concept regardless of the different instance_token used. File names are like `photo of woman.png`. Based on Preview Prompts, the class_token woman is replaced with instance_token + class_token

Example  concepts json file: 
[
    {
        "instance_prompt": "[filewords]",
        "class_prompt": "photo of a woman",
        "save_sample_prompt": "photo of aaa woman",
        "instance_data_dir": "D:\\dev\\images\\aaa",
        "class_data_dir": "D:\\dev\\images\\classification\\woman",
        "instance_token": "aaa",
        "class_token": "woman"
    },
    {
        "instance_prompt": "[filewords]",
        "class_prompt": "photo of a woman",
        "save_sample_prompt": "photo of bbb woman",
        "instance_data_dir": "D:\\dev\\images\\bbb",
        "class_data_dir": "D:\\dev\\images\\classification\\woman",
        "instance_token": "bbb",
        "class_token": "woman"
    },
]

**Comment by Pychnight** [comment]
Just to note, this issue has taken root after commit 4ca69a904f5ddd5651d87032b3dca515eea505ba.

I have been investigating multiple issues related to concept training and concept lists that have occurred recently in the most recent commits of Dreambooth. The issues occur due to a lack of classification data being generated and some internal problems with the code. With the recent commits, it seems the required concepts for data regulation are not being generated correctly (such as generating 5 out of 90 images). As a result, all required concept data cannot be generated. If you use the commit 4ca69a904f5ddd5651d87032b3dca515eea505ba, you will notice it will generate the required class data and will adapt better to the training data.

Commit 5c186f3724614aced14127ef2c66027eb499fff0 is likely the cause of some of the concept generation and generation problems. I have fixed quite a few major regressions in my local branch, but I am still gathering data and creating unit tests to ensure proper user testing is done automatically to test each feature for every change. However, I cannot make any promises.

I have been finding more strange problems that were not happening 1-3 weeks ago, such as concepts not training properly or adapting to the images provided for the dataset. I am creating more test data to narrow down the issues and eventually fix the problems once I can reproduce them and then attach a debugger or memory watcher to fully understand what exactly is happening within the code. This will provide people with enough data to produce a fix.

I have mostly narrowed it down to a few specific changes within the main branch recently that have caused problems. I am just running tests to ensure that the data is correct. As with any user testing or bug finding, solid QA and unit testing are a must.

**Comment by tsngo** [comment]
I was using two \\, it just got formatted a single one. For the arguments I have `--xformers --no-half`, `mixed_percision: no` and `Memory Attention: default` I also tried `mixed_percision: fp16` and `Memory Attention: xformers` but result is same

  ***** Running training *****
  Instance Images: 116
  Class Images: 8869
  Total Examples: 232
  Num batches each epoch = 116
  Num Epochs = 100
  Batch Size Per Device = 1
  Gradient Accumulation steps = 1
  Total train batch size (w. parallel, distributed & accumulation) = 1
  Text Encoder Epochs: 100
  Total optimization steps = 11600
  Total training steps = 11600
  Resuming from checkpoint: False
  First resume epoch: 0
  First resume step: 0
  Lora: False, Adam: True, Prec: no
  Gradient Checkpointing: True
  EMA: True
  LR: 1e-06)

**Comment by taleksan** [comment]
Just wanted to quickly confirm that syncing to an older checkpoint, like 681b953d2f658ef79818b4d2959235ceeeffaf3a (from December) fixes all the above issues for me.

**Comment by tsngo** [comment]
Last good one for me is https://github.com/d8ahazard/sd_dreambooth_extension/commit/c5cb58328c555ac27679422b1da940a9b19de6f2

**Comment by xjdeng** [comment]
I'm not sure if this is still ongoing but I noticed a subtle but related issue: when I train with 2 or more concepts, the images that my model generates looks a lot like the 2nd concept than the 1st.  I can still get the 1st concept if I prompt for it but looking at the source code briefly, I cannot find any place where the training images are shuffled.  If that's the case, then I recommend that the order of the images be shuffled every time.
